jobsdb/tx.go

```diff
@@ -0,0 +1,106 @@
+package jobsdb
+
+import (
+	"database/sql"
+	"errors"
+	"sync"
+)
+
+// ErrEmptyxTx sentinel error indicating an Empty Tx
+var ErrEmptyxTx = errors.New("jobsdb: empty tx")
+
+// TxGetter provides access to an sql.Tx
+type TxGetter interface {
+	// Tx retrieves the tx or returns an error if the tx could not be retrieved (started)
+	Tx() (*sql.Tx, error)
+	// MustTx retrieves the tx or panics in case of an error
+	MustTx() *sql.Tx
+}
+
+type noTx struct{}
+
+// EmptyTx returns an empty interface usable only for tests
+func EmptyTx() TxGetter {
+	return &noTx{}
+}
+
+func (*noTx) Tx() (*sql.Tx, error) {
+	return nil, ErrEmptyxTx
+}
+
+func (*noTx) MustTx() *sql.Tx {
+	return nil
+}
+
+type eagerTx struct {
+	tx *sql.Tx
+}
+
+func (r *eagerTx) Tx() (*sql.Tx, error) {
+	return r.tx, nil
+}
+
+func (r *eagerTx) MustTx() *sql.Tx {
+	return r.tx
+}
+
+type lazyTx struct {
+	dbHandle *sql.DB
+	once     sync.Once
+	tx       *sql.Tx
+	err      error
+}
+
+func (r *lazyTx) Tx() (*sql.Tx, error) {
+	r.once.Do(func() {
+		r.tx, r.err = r.dbHandle.Begin()
+	})
+	return r.tx, r.err
+}
+
+func (r *lazyTx) MustTx() *sql.Tx {
+	tx, err := r.Tx()
+	if err != nil {
+		panic(err)
+	}
+	return tx
+}
+
+// StoreSafeTx sealed interface
+type StoreSafeTx interface {
```

Comment:
It is not clear to me what a "store safe" transaction is. Can you explain?
If I look at the (huge) `JobsDB` interface it says:

```go
	// WithStoreSafeTx prepares a store-safe environment and then starts a transaction
	// that can be used by the provided function.
```
What is a store-safe environment?

Reply:
it makes sure all necessary locks have been acquired